src/java/picard/sam/FixNmAndUqTags.java

```diff
@@ -0,0 +1,123 @@
+        usage = FixNmAndUqTags.USAGE_SUMMARY + FixNmAndUqTags.USAGE_DETAILS,
+        usageShort = FixNmAndUqTags.USAGE_SUMMARY,
+        programGroup = SamOrBam.class
+)
+public class FixNmAndUqTags extends CommandLineProgram {
+    static final String USAGE_SUMMARY = "Fixes the UQ and NM tags in a SAM file.  ";
+    static final String USAGE_DETAILS = "This tool takes in a SAM or BAM file (sorted by coordinate) and calculates the NM and UQ tags by comparing with the reference."+
+            "<br />" +
+                    "This may be needed when MergeBamAlignment was run with SORT_ORDER different from 'coordinate' and thus could not fix\n"+
+                    "these tags then.<br />"+
+            "<h4>Usage example:</h4>" +
+            "<pre>" +
+            "java -jar picard.jar FixNmAndUqTags \\<br />" +
+            "      I=sorted.bam \\<br />" +
+            "      O=fixed.bam \\<br />"+
+            "</pre>" +
+            "<hr />";
+    @Option(doc = "The BAM or SAM file to fix.", shortName = StandardOptionDefinitions.INPUT_SHORT_NAME)
+    public File INPUT;
+
+    @Option(doc = "The fixed BAM or SAM output file. ", shortName = StandardOptionDefinitions.OUTPUT_SHORT_NAME)
+    public File OUTPUT;
+
+    @Option(doc = "If true, assume that the input files are in the same sort order as the requested output sort order, even if their headers say otherwise.",
+            shortName = StandardOptionDefinitions.ASSUME_SORTED_SHORT_NAME)
+    public boolean ASSUME_SORTED = false;
+
+    @Option(doc = "Whether the file contains bisulfite sequence (used when calculating the NM tag).")
+    public boolean IS_BISULFITE_SEQUENCE = false;
+
+    @Override
+    protected String[] customCommandLineValidation() {
+        if (REFERENCE_SEQUENCE == null)
+            return new String[]
+                    {"Must have a non-null REFERENCE_SEQUENCE"};
+        return super.customCommandLineValidation();
+    }
+
+    private final Log log = Log.getInstance(FixNmAndUqTags.class);
+
+    public static void main(final String[] argv) {
+        new FixNmAndUqTags().instanceMainWithExit(argv);
+    }
+
+    protected int doWork() {
+        IOUtil.assertFileIsReadable(INPUT);
+        IOUtil.assertFileIsWritable(OUTPUT);
+        final SamReader reader = SamReaderFactory.makeDefault().referenceSequence(REFERENCE_SEQUENCE).open(INPUT);
+
+        if (!ASSUME_SORTED && reader.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
+            throw new SAMException("Input must be coordinate-sorted, or ASSUME_SORTED must be true.");
+        }
+        if (ASSUME_SORTED && reader.getFileHeader().getSortOrder() != SAMFileHeader.SortOrder.coordinate) {
+            log.warn("File indicates sort-order " + reader.getFileHeader().getSortOrder() + " but ASSUME_SORTED is true, so continuing.");
+        }
+
+        final SAMFileWriter writer = new SAMFileWriterFactory().makeSAMOrBAMWriter(reader.getFileHeader(), true, OUTPUT);
```

Comment:
why? if the input file is valid, I'm generating a file that has the same order and the same header...